cairo/scripts/zkpi_to_stf.py

```diff
@@ -0,0 +1,566 @@
+                        value=hex_to_u256(tx["value"]),
+                        data=Bytes(hex_to_bytes(tx["data"])),
+                        v=hex_to_u256(tx["v"]),
+                        r=hex_to_u256(tx["r"]),
+                        s=hex_to_u256(tx["s"]),
+                    )
+                    if isinstance(tx, dict)
+                    else Bytes(hex_to_bytes(tx))
+                )  # Non-legacy txs are hex strings
+                for tx in encoded_transactions
+            ),
+            ommers=(),
+            withdrawals=tuple(
+                Withdrawal(
+                    index=U64(int(w["index"], 16)),
+                    validator_index=U64(int(w["validatorIndex"], 16)),
+                    address=Address(hex_to_bytes(w["address"])),
+                    amount=U256(int(w["amount"], 16)),
+                )
+                for w in block["withdrawals"]
+            ),
+        )
+
+        output = apply_body(
+            state=blockchain.state,
+            block_hashes=get_last_256_block_hashes(blockchain),
+            coinbase=block.header.coinbase,
+            block_number=block.header.number,
+            base_fee_per_gas=block.header.base_fee_per_gas,
+            block_gas_limit=block.header.gas_limit,
+            block_time=block.header.timestamp,
+            prev_randao=block.header.prev_randao,
+            transactions=block.transactions,
+            chain_id=blockchain.chain_id,
+            withdrawals=block.withdrawals,
+            parent_beacon_block_root=block.header.parent_beacon_block_root,
+            excess_blob_gas=calculate_excess_blob_gas(blockchain.blocks[-1].header),
+        )
+
+        ## Sanity checks against real data
+        assert (
+            output.receipt_root.hex() == data["blocks"][0]["header"]["receiptsRoot"][2:]
+        )
+        assert (
+            output.transactions_root.hex()
+            == data["blocks"][0]["header"]["transactionsRoot"][2:]
+        )
+        assert (
+            output.withdrawals_root.hex()
+            == data["blocks"][0]["header"]["withdrawalsRoot"][2:]
+        )
+        assert (
+            output.block_logs_bloom.hex()
+            == data["blocks"][0]["header"]["logsBloom"][2:]
+        )
+        assert output.block_gas_used == Uint(
+            int(data["blocks"][0]["header"]["gasUsed"][2:], 16)
+        )
+        assert output.blob_gas_used == Uint(
+            int(data["blocks"][0]["header"]["blobGasUsed"][2:], 16)
+        )
+
+        post_state = blockchain.state
+        state_diff = StateDiff.from_pre_post(pre_state_copy, post_state)
+        ethereum_state.update_from_state_diff(state_diff)
+        success = (
+            ethereum_state.state_root.hex()
+            == data["blocks"][0]["header"]["stateRoot"][2:]
+        )
+        assert (
+            ethereum_state.state_root.hex()
+            == data["blocks"][0]["header"]["stateRoot"][2:]
+        )
```

Comment:
I'd rather use `state_transition` directly, with a patched `state_root`. At least make sure to reproduce exactly its body, with a comment

Reply:
proposing something in my next commit